server/apps/main/views.py

```diff
@@ -523,82 +524,113 @@ def moderate_experience(request, uuid):
         model = PublicExperience.objects.get(experience_id=uuid)
         if request.method == "POST":
             # get the data from the model
-            moderated_form = ModerateExperienceForm(request.POST)
-            moderated_form.is_valid()
-
-            # Get the (immutable) experience data
-            unchanged_experience_details = extract_experience_details(model)
-            # Get the (mutable) trigger warnings
-            trigger_details = process_trigger_warnings(moderated_form)
-
-            # validate
-            data = {**unchanged_experience_details, **trigger_details}
+            form = ModerateExperienceForm(request.POST)
+            if form.is_valid():
+                # Get the (immutable) experience data
+                unchanged_experience_details = extract_experience_details(model)
+                # Get the (mutable) trigger warnings
+                trigger_details = process_trigger_warnings(form)
+
+                # validate
+                data = {**unchanged_experience_details, **trigger_details}
+
+                moderation_comments = data.pop("moderation_comments", None)
+                moderation_reply = data.pop("moderation_reply", "")
+                moderation_prior = data.pop("moderation_prior", "not moderated")
+
+                # get the users OH member id from the model
+                user_OH_member = model.open_humans_member
+
+                # wrap in try/except in case user writing experience originally has left AutSPACEs
+                try:
+                    user_OH_member.delete_single_file(file_basename=f"{uuid}.json")
+                    upload(data, uuid, ohmember=user_OH_member)
+                    # update the PE object
+                    update_public_experience_db(
+                        data,
+                        uuid,
+                        ohmember=user_OH_member,
+                        editing_user=request.user.openhumansmember,
+                        change_type="Moderate",
+                        change_comments=moderation_comments,
+                        change_reply=moderation_reply,
+                    )
+                except Exception:
+                    # if user writing E has deauthorized autspaces, delete public experience
+                    delete_PE(uuid=uuid, ohmember=user_OH_member)
+
+                # send a message to the author if they've requested it
+                moderation_status = data.get("moderation_status", "")
+                moderation_string = moderation_status.title()
+                try:
```

Comment:
I'm not 100% sure why we have to wrap this in `try/expect`, is it in case the `UserProfile` doesn't exist? Just trying to better understand the edge cases we're trying to catch here!

Reply:
There are two possible exceptions that I'm aware of potentially being caught here. One of them, as you mention, is a potentially missing `UserProfile`. The other is in case reading of the `mod_message.txt` file fails for some reason, in which case an exception may bubble up from `get_moderator_message()`.

We could catch them separately to make this more explicit (or remove the `UserProfile` catch completely?) if you prefer?

Reply:
Given that the lack of `UserProfile` also creates the bug in #552, we maybe should make a helper to check if the user has a profile and that returns `None` (or similar) otherwise to help us check those things in conditionals without also having to manually check the presence of the `UserProfile`? As I think for this one we will have to keep checking its existence to make sure we can send a message. 

For the `mod_message.txt`, I guess given that our _default_ is part of the repo we could assume that it exists? :)

Reply:
Yeah, that makes sense to me!

Reply:
There's now a helper function for safely getting the user profile. I've used it here in the moderation view, but the #552 task is still needed for the other cases.

Reply:
Perfect!